proof/refine/ARM/Schedule_R.thy

```diff
@@ -4210,6 +4209,1003 @@ lemma refillUnblockCheck_corres':
+                                       (sc_refill_cfgs_of s) scPtr
+                           \<and> scPtr = cur_sc s"
+              and Q'="\<lambda>_ s'. valid_refills' scPtr s' \<and> active_sc_at' scPtr s' \<and> scPtr = ksCurSc s'"
+              and r'=dc
+               in corres_split'[rotated])
+     apply (corressimp corres: headInsufficientLoop_corres)
+     apply (fastforce simp: vs_all_heap_simps word_le_nat_alt)
+    apply (intro hoare_vcg_conj_lift_pre_fix; (solves wpsimp)?)
+       apply schedule_used_simple
+       apply (clarsimp simp: obj_at_def is_sc_obj_def)
+      apply schedule_used_simple
+     apply (wpsimp wp: schedule_used_refills_unat_sum update_sched_context_wp
+                 simp: update_refill_hd_def)
+     apply (clarsimp simp: obj_at_def vs_all_heap_simps refills_unat_sum_cons
+                           refills_unat_sum_append)
+     apply (subst unat_sub)
+      apply fastforce
+     apply (clarsimp simp: word_less_nat_alt)
+     apply (drule less_imp_le)
+     apply (clarsimp simp: refills_unat_sum_def)
+     apply (case_tac "sc_refills sc"; clarsimp simp: refills_unat_sum_cons)
+    apply schedule_used_simple
+   apply wpsimp
+
+  apply (rule_tac F="refill_hd sc = refill_map (refillHd sc')" in corres_req)
+   apply clarsimp
+   apply (rule refill_hd_relation)
+     apply fastforce
+    apply (clarsimp simp: vs_all_heap_simps obj_at_def)
+   apply (fastforce simp: valid_refills'_def obj_at_simps opt_map_red)
+
+  apply (clarsimp simp: maxReleaseTime_equiv)
+  apply (simp add: when_def split del: if_split)
+  apply (rule corres_if_split; (solves simp)?)
+   apply (clarsimp simp: refill_map_def)
+  apply (rule corres_symb_exec_l[rotated 2, OF get_sched_context_sp])
+    apply wpsimp
+    apply (clarsimp simp: obj_at_def)
+   apply (find_goal \<open>match conclusion in "\<lbrace>P\<rbrace> f \<exists>\<lbrace>Q\<rbrace>" for P f Q \<Rightarrow> -\<close>)
+   apply (wpsimp wp: get_sched_context_exs_valid)
+    apply (clarsimp simp: obj_at_def)
+   apply simp
+  apply (rename_tac new_sc)
+  apply (rule_tac F="new_sc=sc" in corres_req)
+   apply (clarsimp simp: obj_at_def)
+  apply (subst update_refill_hd_rewrite)
+  apply (clarsimp simp: bind_assoc)
+  apply (rule corres_guard_imp)
+    apply (rule corres_split[OF updateRefillHd_corres])
+        apply simp
+       apply (clarsimp simp: refill_map_def)
+      apply (rule corres_split[OF updateRefillHd_corres])
+          apply simp
+         apply (clarsimp simp: refill_map_def)
+        apply (rule scheduleUsed_corres)
+         apply simp
+        apply (clarsimp simp: refill_map_def sc_relation_def)
+       apply wpsimp
+      apply (wpsimp wp: updateRefillHd_valid_objs')
+     apply wpsimp
+    apply (wpsimp wp: updateRefillHd_valid_objs')
+   apply (clarsimp simp: is_active_sc_rewrite)
+  apply (fastforce intro: valid_objs'_valid_refills'
+                    simp: active_sc_at'_def is_active_sc'_def obj_at_simps opt_map_red)
```

Comment:
Does `valid_objs’` actually show up in the proofs here? If so, where does it come from I wonder?

Reply:
Removed